velox/exec/HashBuild.h

```diff
@@ -18,35 +18,103 @@
+  /// nothing. In this case, HashBuild operators finishes early without
+  /// processing all the input and without finishing building the hash table.
   struct HashBuildResult {
+    HashBuildResult(
+        std::shared_ptr<BaseHashTable> _table,
+        std::optional<SpillPartitionId> _restoredPartitionId,
+        SpillPartitionIdSet _spillPartitionIds)
+        : antiJoinHasNullKeys(false),
+          table(std::move(_table)),
+          restoredPartitionId(std::move(_restoredPartitionId)),
+          spillPartitionIds(std::move(_spillPartitionIds)) {}
+
+    HashBuildResult() : antiJoinHasNullKeys(true) {}
+
+    bool antiJoinHasNullKeys{false};
     std::shared_ptr<BaseHashTable> table;
-    bool antiJoinHasNullKeys;
+    std::optional<SpillPartitionId> restoredPartitionId;
+    SpillPartitionIdSet spillPartitionIds;
+  };
+
+  std::optional<HashBuildResult> tableOrFuture(
+      ContinueFuture* FOLLY_NONNULL future);
+
+  /// Invoked by HashProbe operator after finishes probing the built table to
+  /// set one of the previously spilled partition to restore. The HashBuild
+  /// operators will then build the next hash table from the selected spilled
+  /// one.
+  bool setSpillInput();
+
+  /// Contains the spill input for one HashBuild operator: a shard of previously
+  /// spilled partition data. 'spillPartition' is null if there is no more spill
+  /// data to restore.
+  struct SpillInput {
+    explicit SpillInput(
+        std::unique_ptr<SpillPartition> spillPartition = nullptr)
+        : spillPartition(std::move(spillPartition)) {}
+
+    std::unique_ptr<SpillPartition> spillPartition;
   };
 
-  std::optional<HashBuildResult> tableOrFuture(ContinueFuture* future);
+  /// Invoked by HashBuild operator to get one of previously spilled partition
+  /// shard to restore. The spilled partition to restore is set by HashProbe
+  /// operator after finishes probing on the previously built hash table.
+  /// If HashBuild operator calls this early, 'future' will be set to wait
+  /// asynchronously. If there is no more spill data to restore, then
+  /// 'spillPartition' will be set to null in the returned SpillInput.
+  std::optional<SpillInput> spillInputOrFuture(
+      ContinueFuture* FOLLY_NONNULL future);
 
  private:
-  std::shared_ptr<BaseHashTable> table_;
-  bool antiJoinHasNullKeys_{false};
+  uint32_t numBuilders_{0};
+
+  std::optional<HashBuildResult> buildResult_;
+
+  // If not null, set to the currently restoring spill partition id.
+  std::optional<SpillPartitionId> restoringSpillPartitionId_;
+
+  // If 'restoringSpillPartitionId_' is not null, this set to the restoring
+  // spill partition data shards. Each shard is expected to have the same number
+  // of spill files and will be processed by one of the HashBuild operator.
+  std::vector<std::unique_ptr<SpillPartition>> restoringSpillShards_;
+
+  // The spill partitions remaining to restore.
+  SpillPartitionSet spillPartitionSets_;
```

Comment:
Just to clarify this set is populated using information provided by the HashBuild operators. This set can grow if HashBuild operator cannot load full partition in memory and engages in recursive spilling.

Reply:
Add this into comment. Thanks!